### DISCUSSION TITLE
Community Made MIDI Packs for WADFusion

### DISCUSSION THREAD

**Discussion by Ohthatguy43** [Owlet7/wadfusion#29] [Show and tell]
An update to an old mod for the original WadSmoosh, this mod adds in community midi packs for:

Blackroom Warmup Levels (Ultimate MIDI Pack)

Sewers and Betray (Ultimate MIDI Pack and .MID the way ID did)

Ultimate DOOM Episode 4 (Ultimate MIDI Pack)

DOOM II (.MID the way ID did)

FINAL Doom (TNT: Evilution MIDI Pack and Plutonia MIDI Pack)

No Rest For The Living (NERVE MIDI Pack)

The Master Levels + Rejects (Master Levels MIDI Pack for all final master levels and Ultimate MIDI Pack, and unused DOOM music for rejected levels.)

There is also an option to swap between jaguar levels for E3M1 and E2M9 respectively.

As a bonus, The Master Levels include their own intermission graphic!

Special thanks for Owlet7 and domyoji81 for fixing my janky ass mod

Download for stable builds: https://github.com/user-attachments/files/19103053/wadfusion_midipacks.zip

Download for DEV versions of WADFusion: https://drive.google.com/file/d/1CuwEhQTTbn77eVNNiKczTOl5Rgt1-dha/view?usp=sharing

**Comment by Ohthatguy43** [comment]
Lil update, But i added Jaguar DOOM's Tower of Babel (Hell Gate) as a map replacement for e2m8 (using code from the blackroom stuff) and tried to do the same for Jaguar's Version of Hell Keep (Hell's Keep) though since i'm not a coder at all, i'm not able to figure out how to do it for e3m1.

Download: https://drive.google.com/file/d/1C8bsJkpGv0YX4TPbSTgTLcGmAmuRF4cn/view?usp=sharing

**Reply by Owlet7** [reply]
This update breaks WF's EventHandler, you replaced it with a new one with the same name.

This block of code doesn't do what you want it to:
```
if ( mapName == "E2M9" )
{
	if ( !CVar.FindCVar("wf_jaguar_e3m1b").GetBool() )
		level.NextMap = "e3m1";
	else
		level.NextMap = "e3m1b";
}
```
This makes it so that the secret map of episode 2 will lead directly into episode 3.

**Comment by Ohthatguy43** [comment]
Fixed, sort of.
I still have no clue how to get the E3M1 replacement to work properly currently but the handler stuff should be fixed
https://drive.google.com/file/d/1C8bsJkpGv0YX4TPbSTgTLcGmAmuRF4cn/view?usp=sharing same link

**Reply by domyoji81** [reply]
> Fixed, sort of. I still have no clue how to get the E3M1 replacement to work properly currently but the handler stuff should be fixed https://drive.google.com/file/d/1C8bsJkpGv0YX4TPbSTgTLcGmAmuRF4cn/view?usp=sharing same link

You could explore using ChangeLevel to accomplish this. Something along the lines of this under WorldTick:

```
String mapName = level.MapName.MakeLower();
if ( mapName == "e3m1" )
     {
          if (CVar.FindCVar("your_bool_name").GetBool() )
               Level.ChangeLevel("e3m1b",0,CHANGELEVEL_NOINTERMISSION,-1);
     }
```

It's not a perfectly ideal solution by itself because it'll still show e3m1 for a tick (and the subsequent wipe to e3m1b), but you could hide that behind a blank/black screen with a Draw or Overlay. There might be a more elegant solution though, I don't know.

**Comment by Ohthatguy43** [comment]
wow Jolly update!
changelog:
Hell Gate no longer replaces E2M8, it now replaces E2M9 as i think it much better fits that level
Updated mapinfo a lil to sync with current wadfusion changes
Added SIGIL + SIGIL II widescreen assets (Again, made by Scuba Steve).
Added .MID the way ID did songs for DOOM II (idk why i didn't do that)
Replaced all .MID the way ID did that were used for master levels rejects with rejected doom songs
that's it i think Lol
Download: https://drive.google.com/file/d/1C8bsJkpGv0YX4TPbSTgTLcGmAmuRF4cn/view?usp=sharing

**Reply by Someguyman** [reply]
Honestly, I would have replaced e3m9 with hell gate since the whole reason warrens exists is that it parallels the first map of the hell levels and having the console version replace the first level devalues warrens as the secret map. So replacing it with hell gate would address the issue.

**Comment by Ohthatguy43** [comment]
Very quick update but i fixed some incompitiblity issues when using 1.3 wadfusion, that's it really
Download: https://drive.google.com/file/d/1C8bsJkpGv0YX4TPbSTgTLcGmAmuRF4cn/view?usp=sharing

**Reply by 321gametime** [reply]
Seems like the new update broke compatibility. I also noticed in the last version that the Midi pack for Plutonia wouldn't work if the IDKFA soundtrack was turned on, but the Midi packs for the other episodes worked fine with it on.  Would like a fix, but it's honestly not a big deal. I'm just happy this exists in the first place.

**Reply by Ohthatguy43** [reply]
Thanks for letting me know, i will update this mod in the next couple of days probably

**Comment by Ohthatguy43** [comment]
New update. fixes incompatiblity with wadfusion 1.4.1
download: https://drive.google.com/file/d/bc1qpjq2u4vwnyxe3u0flg8hfp3c8mfmd9f38ydu07/view?usp=sharing
(PS: I couldn't figure out the bug related to IDKFA ost and plutonia, sorry!)

**Reply by Owlet7** [reply]
The link appears to be incorrect.
> Sorry, the file you have requested does not exist.
> Make sure that you have the correct URL and that the file exists.

**Reply by Ohthatguy43** [reply]
Please try this: https://drive.google.com/file/d/1C8bsJkpGv0YX4TPbSTgTLcGmAmuRF4cn/view?usp=sharing

**Reply by 321gametime** [reply]
Thank link works and the mod itself is working just fine. Thanks.

**Reply by Owlet7** [reply]
Here's a fixed version that hopefully should have less compatibility issues with future WF patches:
[wadfusion_midipacks.zip](https://github.com/user-attachments/files/19103053/wadfusion_midipacks.zip)

I took the liberty of also fixing a broken texture, and removing the Sigil widescreen graphics which ~~will become obsolete in the next patch~~ are now obsolete.

I also improved the level transitions for the replaced maps, but I suggest taking @Someguyman's suggestion to replace E3M9 instead of E3M1. His logic seems sound.

The IDKFA music in Plutonia issue was actually a bug in Fusion, ~~it'll also be fixed in the next patch~~ it's fixed now.

**Reply by Owlet7** [reply]
By the way, I suggest also posting updates to the OP so people don't accidentally download an old version.

**Reply by 321gametime** [reply]
> I took the liberty of also fixing a broken texture, and removing the Sigil widescreen graphics which will become obsolete in the next patch are now obsolete.

So does this mean there is no reason to use the kex wads over the originals, since the widescreen assets are going to come from this mod?

> The IDKFA music in Plutonia issue was actually a bug in Fusion, it'll also be fixed in the next patch it's fixed now.

I'm still having problems with Plutonia and IDKFA. Maybe this is an issue on my end? I tried just loading Wadfusion and the midipack, but it's still playing the wrong music when I have IDKFA selected.

**Reply by Owlet7** [reply]
> So does this mean there is no reason to use the kex wads over the originals, since the widescreen assets are going to come from this mod?

Well, that's down to the user's own preference... But yes.

> I'm still having problems with Plutonia and IDKFA. Maybe this is an issue on my end? I tried just loading Wadfusion and the midipack, but it's still playing the wrong music when I have IDKFA selected.

By "playing the wrong music" do you mean that it plays the IDKFA soundtrack when it's turned on? Because that's the intended working behaviour. Or do you mean that it plays the wrong music when the IDKFA music is switched back off?

**Reply by 321gametime** [reply]
> By "playing the wrong music" do you mean that it plays the IDKFA soundtrack when it's turned on? Because that's the intended working behaviour. Or do you mean that it plays the wrong music when the IDKFA music is switched back off?

When IDKFA is on, the midi pack still plays for the other episodes, like Thy Flesh Consumed and the master levels. Plutonia is the only one that plays IDKFA when it is selected. It's a minor thing, but I think would be cool for it to also prioritize the midi pack over IDKFA.

**Reply by Ohthatguy43** [reply]
> By the way, I suggest also posting updates to the OP so people don't accidentally download an old version.

It has been updated 👍

**Reply by Owlet7** [reply]
>When IDKFA is on, the midi pack still plays for the other episodes, like Thy Flesh Consumed and the master levels. Plutonia is the only one that plays IDKFA when it is selected.

That's incorrect. Try playing E2M3.

> It's a minor thing, but I think would be cool for it to also prioritize the midi pack over IDKFA.

The mod (the MIDI Pack) changes the music files used in some levels to ones that have filenames that the IDKFA script doesn't account for. That's why the MIDI pack "gets prioritised" on those levels. Other levels, like Doom 1, Doom 2, and Final Doom (both TNT and Plutonia) replace the "duplicate" music with files that use the original filenames, so they get replaced by the IDKFA script.

This is the intended and desired behaviour for WadFusion, but the mod can remedy this by giving every new music track a unique filename.

**Reply by 321gametime** [reply]
> The mod (the MIDI Pack) changes the music files used in some levels to ones that have filenames that the IDKFA script doesn't account for. That's why the MIDI pack "gets prioritised" on those levels. Other levels, like Doom 1, Doom 2, and Final Doom (both TNT and Plutonia) replace the "duplicate" music with files that use the original filenames, so they get replaced by the IDKFA script.

Ok that makes sense. I didn't try those levels yet. I just assumed that because Thy Flesh Consumed worked Doom 1 as a whole did. Sorry for the confusion.

**Comment by Ohthatguy43** [comment]
New version of this mod for the current dev builds of wadfusion, it's a seperate download that doesnt replace the original one, here are the changes:


Added intro and read intermission songs to the master levels and nerve
fixed issue with nerve intermission bgs not loading
synced mapinfo changes from dev to here
hell gate now replaces e3m9 and not e2m9, as suggested by @Someguyman 
thats about it i think
download: https://drive.google.com/file/d/1CuwEhQTTbn77eVNNiKczTOl5Rgt1-dha/view?usp=sharing

**Reply by PotatoMaster-007** [reply]
Hi, I wanted to try opening the mod in the current dev version (from uzdoom), but I'm getting this message:

Script error, "doom_fusion.ipk3:zscript/wf_handler_static.zs" line 58: Can't call ui function NewGameChangeLevelInput from play context (not readable)

I thought it might be related to the uzdoom dev version, but when I tried wadfusion xtras, it ran without any problems.

**Reply by Owlet7** [reply]
The issue appears to be that this mod replaces a Fusion dev build script (`wf_newgame.zs`) with an older copy of itself.

**Reply by Ohthatguy43** [reply]
> The issue appears to be that this mod replaces a Fusion dev build script (`wf_newgame.zs`) with an older copy of itself.

+ This being a couple months old when compared to the latest one now, i hope to get around to updating it when 1.6.0 of wadfusion releases, eventually.

keyword eventually

**Reply by Owlet7** [reply]
Why did you need to include it, anyway? To set the title music for some of the episodes? You don't need to copy everything for that.

**Reply by Ohthatguy43** [reply]
> Why did you need to include it, anyway? To set the title music for some of the episodes? You don't need to copy everything for that.

not sure how else i can do it considering im a noob at zscript so this was the ""best"" solution

**Reply by 321gametime** [reply]
I didn't know there was a dev build version of wadfusion. Is it different from the release on the github?

**Reply by Owlet7** [reply]
It's the master branch on here itself. I consider any changes to it between releases to be "dev build" versions.

**Comment by 321gametime** [comment]
Trying to get this to work on the new UZDoom source port and I'm getting this error. 

> Execution could not continue.
> Script error, "wadfusion_midipacks.zip:zscript.zs" line 21:
> The file you are attempting to run requires a newer version of UZDoom.
> 
> A version with ZScript version 4.15.0 is required, but your copy of UZDoom only supports 4.14.3. Please upgrade!

Not sure if this is as problem with the mod itself of (most likely) the source port. Or maybe I'm being dumb and I'm doing something wrong lol.

**Reply by Owlet7** [reply]
~~The mod copies some code from Fusion that requires using dev builds of ZDoom.~~ The UZDoom 4.14.3 release candidate can be used with WadFusion 1.5.1, but newer changes to Fusion require features that won't make it into this release of UZDoom. Your options are either to use an older version of this mod, or to use dev builds of both UZDoom and WadFusion.

EDIT: The mod doesn't actually need to copy this code, it can be updated to fix this.

**Comment by james1222312334** [comment]
btw, can you make a version that works with UZDoom 4.13.3
